### ISSUE TITLE
Some basic example with Kernel

### ISSUE THREAD

**Issue by kolkov** [chewxy/cu#21]
Hi! 
Please provide some basic example how to work with this library properly.
How to create basic Kernel and use it into computation with GPU.

Thanks!

**Comment by chewxy** [comment]
Yeah I'll get round to doing that.

**Comment by kolkov** [comment]
It is better to give two examples. One is quite simple, to understand the basics, and one more complex, to understand where to move on.

**Comment by cobnst** [comment]
@chewxy @kolkov well  how to found the example? i'm miss 0.0

**Comment by Icaro-Lima** [comment]
I'm here 2 years later ... Any examples?

**Comment by malkhamis** [comment]
maybe [this test file](https://github.com/gorgonia/cu/blob/master/batch_test.go) can serve as an example :thinking:  ?

**Comment by Icaro-Lima** [comment]
This is exactly what I need!

**Comment by neurlang** [comment]
I got a [working kernel here](https://github.com/neurlang/classifier/blob/master/learning/cu/learn.go)

tl;dr: The order of operations which seems to be working is:
**initCUDA()**
```
device, err := cu.GetDevice(0)
ctx, err := device.MakeContext(cu.SchedAuto)
err = ctx.Lock()
// then do malloc etc for permanent memory
mod, err := cu.LoadData(kernel.PTXreduceCUDA)
fn, err := mod.Function("reduce")
stream, err := cu.MakeStream(cu.DefaultStream)
```
**reduceCUDA()**
```
err := cu.SetCurrentContext(ctx)
// single use memory malloc/ permanent memory clear (memset 0)
err = cu.MemcpyHtoD(d_input_nums, unsafe.Pointer(&input_numbers[0]), inputNumsSize) // copy some data to kernel
err = d_fn.LaunchAndSync(x[1][0], x[1][1], x[1][2], x[0][0], x[0][1], x[0][2], 0, d_stream, args)
err = cu.MemcpyDtoH(unsafe.Pointer(&result[0]), d_result, resultSize) // copy solution data back from device to host
```

**destroyCUDA()**
```
cu.MemFree(memory) // free all memories
ctx.Unlock()
ctx.Destroy()
```

the kernel itself is in the kernel/ subdir, compiled to .ptx as this:
```
#!/bin/bash
nvcc -ptx reduceCUDA.cu -o reduceCUDA.ptx
```
the kernel does have a sort of main function:
```
extern "C" __global__ void reduce(uint8_t *d_set, uint32_t *d_nums, uint32_t *alphabet, uint32_t* out) {
// the code
}
```

Hope this serves as a solid example to the next person.

**Comment by chewxy** [comment]
hey @neurlang do you want to put this in a PR as an example?

**Comment by neurlang** [comment]
pushing hashtron kernel to the main cu repo doesn't feel right. it would shift the maintenance burden to you, i still need to tweak the kernel sometimes, etc. I think I will keep it in my repo instead. Anyone can look there anyway.

Speaking of this repo (gorgonia/cu), **I would prefer to put here more testcases which stress the code paths related to my kernel**, to make sure that hashtron behavior is not broken here accidentally by you or by someone, now that it works pretty stable.

On the other hand, now that we (or simply me) know how to make a CU kernel, it would be a good idea to write it to the gorgonia.github.io site as a **tutorial to make a CU kernel**.

I'll open a separate issue on there instead.
